examples/plot_metadata_routing.py

```diff
@@ -0,0 +1,616 @@
+# if the estimator doesn't have such a method, then a default empty
+# ``MetadataRequest`` is returned.
+#
+# Then in each method, we use the ``route_params`` method to construct a
+# dictionary of the form ``{"object_name": {"method_name": {"metadata":
+# value}}}`` to pass to the underlying estimator's method.
+# ``validate_metadata`` makes sure all given metadata are requested. This is to
+# avoid silent bugs, and this is how it will work:
+
+est = MetaClassifier(estimator=ExampleClassifier().set_fit_request(sample_weight=True))
+est.fit(X, y, sample_weight=my_weights)
+
+# %%
+# Note that the above example checks that ``sample_weight`` is correctly passed
+# to ``ExampleClassifier``, or else it would print that ``sample_weight`` is
+# ``None``:
+
+est.fit(X, y)
+
+# %%
+# If we pass an unknown metadata, it will be caught:
+try:
+    est.fit(X, y, test=my_weights)
+except TypeError as e:
+    print(e)
+
+# %%
+# And if we pass something which is not explicitly requested:
+try:
+    est.fit(X, y, sample_weight=my_weights).predict(X, groups=my_groups)
+except ValueError as e:
+    print(e)
+
+# %%
+# Also, if we explicitly say it's not requested, but pass it:
+est = MetaClassifier(
+    estimator=ExampleClassifier()
+    .set_fit_request(sample_weight=True)
+    .set_predict_request(groups=False)
+)
+try:
+    est.fit(X, y, sample_weight=my_weights).predict(X[:3, :], groups=my_groups)
+except TypeError as e:
+    print(e)
+
+# %%
+# Another concept to introduce is aliased metadata. This is when an estimator
+# requests a metadata with a different name than the default value. For
+# instance, in a setting where there are two estimators in a pipeline, one
+# could request ``sample_weight1`` and the other ``sample_weight2``. Note that
+# this doesn't change what the estimator expects, it only tells the
+# meta-estimator how to map provided metadata to what's required. Here's an
+# example, where we pass ``aliased_sample_weight`` to the meta-estimator, but
+# the meta-estimator understands that ``aliased_sample_weight`` is an alias for
+# ``sample_weight``, and passes it as ``sample_weight`` to the underlying
+# estimator:
+est = MetaClassifier(
+    estimator=ExampleClassifier().set_fit_request(sample_weight="aliased_sample_weight")
+)
+est.fit(X, y, aliased_sample_weight=my_weights)
+
+# %%
+# And passing ``sample_weight`` here will fail since it is requested with an
+# alias and ``sample_weight`` with that name is not requested:
+try:
+    est.fit(X, y, sample_weight=my_weights)
+except TypeError as e:
+    print(e)
+
+# %%
+# This leads us to the ``get_metadata_routing``. The way routing works in
+# scikit-learn is that consumers request what they need, and routers pass that
+# along. But another thing a router does, is that it also exposes what it
+# requires so that it can be used inside another router, e.g. a pipeline inside
+# a grid search object. The output of the ``get_metadata_routing`` which is a
```

Comment:
```suggestion
# along. Additionally, a router exposes what it
# requires itself such that it can be used inside another router, e.g. a pipeline inside
# a grid search object. The output of ``get_metadata_routing``, which is a
```